### PR TITLE
feat: 12 add order

### PR DESCRIPTION
**Pull request by EduarfF**

### CODE REVIEW

File: src/main/java/org/example/springbootintro/controller/OrderController.java [hunk 1]
```diff
@@ -0,0 +1,89 @@
...
+    @PreAuthorize("hasRole('USER')")
+    @Operation(
+            summary = "Get all items from current order",
+            description = "Get all items from current order"
+    )
+    public List<OrderItemDto> getAllOrderItemsForSpecificOrder(
+            @PathVariable Long orderId,
+            Pageable pageable
+    ) {
+        return orderItemService.getAll(orderId, pageable);
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/controller/OrderController.java]: Please make sure that one user can not access order data of different user

File: src/main/java/org/example/springbootintro/dto/order/CreateOrderRequestDto.java [hunk 2]
```diff
@@ -0,0 +1,12 @@
+package org.example.springbootintro.dto.order;
+
+import jakarta.validation.constraints.NotBlank;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
+public class CreateOrderRequestDto {
+    @NotNull(message = "Shipping address can not be null")
+    @NotBlank(message = "Shipping address ca not be blank")
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/dto/order/CreateOrderRequestDto.java]: NotBlank already includes NotNull

File: src/main/java/org/example/springbootintro/dto/order/CreateOrderRequestDto.java [hunk 3]
```diff
@@ -0,0 +1,12 @@
...
+    private String shippingAddress;
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/dto/order/CreateOrderRequestDto.java:11]: add length validation to be aligned with db

File: src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java [hunk 4]
```diff
@@ -0,0 +1,101 @@
...
+                user,
+                Status.PENDING,
+                LocalDateTime.now(),
+                requestDto.getShippingAddress()
+        );
+
+        Set<OrderItem> orderItems = getOrderItems(shoppingCart, order);
+
+        order.setTotal(getSumOfAllItems(orderItems));
+        order.setOrderItems(orderItems);
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java]: Move order creation logic to separate method

File: src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java [hunk 5]
```diff
@@ -0,0 +1,101 @@
...
+                    item.getBook(),
+                    item.getQuantity(),
+                    item.getBook().getPrice()
+            );
+            orderItems.add(orderItem);
+        }
+        return orderItems;
+    }
+
+    private static BigDecimal getSumOfAllItems(Set<OrderItem> orderItems) {
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java]:
```suggestion
    private static BigDecimal getTotal(Set<OrderItem> orderItems) {
```

File: src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+        BigDecimal sum = BigDecimal.ZERO;
+        for (OrderItem item : orderItems) {
+            BigDecimal itemTotal = item.getPrice()
+                    .multiply(BigDecimal.valueOf(item.getQuantity()));
+            sum = sum.add(itemTotal);
+        }
+        return sum;
+    }
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java]: Could be done using streams

File: src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+    public OrderDto update(Long id, UpdateOrderItemDto requestDto) {
+        Order order = orderRepository.findById(id)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can not find order with id: " + id)
+                );
+        order.setStatus(requestDto.getStatus());
+        return orderMapper.toDto(orderRepository.save(order));
+    }
+
+    private static Set<OrderItem> getOrderItems(ShoppingCart shoppingCart, Order order) {
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/service/order/OrderServiceImpl.java]:
```suggestion
    private Set<OrderItem> getOrderItems(ShoppingCart shoppingCart, Order order) {
```

File: src/main/java/org/example/springbootintro/service/orderitem/OrderItemServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+@RequiredArgsConstructor
+@Service
+public class OrderItemServiceImpl implements OrderItemService {
+    private final OrderRepository orderRepository;
+    private final OrderItemRepository orderItemRepository;
+    private final OrderItemMapper orderItemMapper;
+
+    @Override
+    public List<OrderItemDto> getAll(Long id, Pageable pageable) {
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/service/orderitem/OrderItemServiceImpl.java]:
```suggestion
    public List<OrderItemDto> getAll(Long orderId, Pageable pageable) {
```

File: src/main/java/org/example/springbootintro/service/orderitem/OrderItemServiceImpl.java [hunk 9]
```diff
@@ -0,0 +1,45 @@
...
+                        "Can not find order with id: " + orderId)
+                );
+        OrderItem orderItem = order.getOrderItems()
+                .stream()
+                .filter(item -> item.getId().equals(itemId))
+                .findFirst()
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can not find order item with id: " + itemId)
+                );
+        return orderItemMapper.toDto(orderItem);
```
Comment (andrii-hoienko) [inline src/main/java/org/example/springbootintro/service/orderitem/OrderItemServiceImpl.java]:
```suggestion
        return orderItemRepo.find(orderId, userId, orderItemId)
```

**Review by andrii-hoienko — CHANGES_REQUESTED** [review]
Nice, left some comments

File: src/main/java/org/example/springbootintro/model/Order.java [hunk 10]
```diff
@@ -0,0 +1,76 @@
...
+@NoArgsConstructor
+@SQLDelete(sql = "UPDATE orders SET is_deleted = true WHERE id=?")
+@SQLRestriction(value = "is_deleted = false")
+@Table(name = "orders")
+public class Order {
+    @Id
+    @GeneratedValue(strategy = GenerationType.IDENTITY)
+    private Long id;
+
+    @ManyToOne
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/model/Order.java:40]: please use LAZY fetch type and it is better to add this even if it's default fetch type for the annotation

File: src/main/java/org/example/springbootintro/service/orderitem/OrderItemServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+                .toList();
+    }
+
+    @Override
+    public OrderItemDto getByItemId(Long userId, Long orderId, Long itemId) {
+        OrderItem orderItem = orderItemRepository
+                .findByOrderIdAndUserIdAndItemId(orderId, userId, itemId)
+                .orElseThrow(() -> new EntityNotFoundException(
+                        "Can not find order item with id: " + itemId)
+                );
```
Comment (yevheniisynytsia) [inline src/main/java/org/example/springbootintro/service/orderitem/OrderItemServiceImpl.java:39]:
```suggestion
                        "Can not find order item with id: " + itemId));
```

**Review by yevheniisynytsia — APPROVED** [review]

**Pull request merged**